### PR TITLE
[FS] Updated sync logic to handle NullReferenceException

### PR DESCRIPTION
**Pull request by fenix2222**
Added a null check to make sure we do not get NullReferenceException, sync should reenter and be ok next time around. Alternatively, we may need to just throw a known exception

### CODE REVIEW

File: src/Stratis.Features.FederatedPeg/TargetChain/CrossChainTransferStore.cs [hunk 1]
```diff
@@ -812,6 +812,13 @@ private bool Synchronize()
             lock (this.lockObj)
             {
                 HashHeightPair tipToChase = this.TipToChase();
+                if (this.TipHashAndHeight == null)
+                {
+                    this.logger.LogError("Failed to synchronise. Reason: TipHashAndHeight is null.");
```
Comment (fassadlr) [inline src/Stratis.Features.FederatedPeg/TargetChain/CrossChainTransferStore.cs]: Change to `this.logger.LogError("Failed to synchronise. Reason: {0} is null.", nameof(TipHashAndHeight));`

**Comment by fenix2222** [comment]
@fassadlr that was to fix race condition, we basically use a snapshot as chainindexer may have changed during sync

File: src/Stratis.Features.FederatedPeg/TargetChain/CrossChainTransferStore.cs [hunk 2]
```diff
@@ -865,21 +872,21 @@ private bool SynchronizeBatch()
...
 
-                blockHashes.Add(header.HashBlock);
+                chainedHeadersSnapshot.Add(header.HashBlock, header);
 
                 if (++batchSize >= synchronizationBatchSize)
                     break;
             }
 
-            List<Block> blocks = this.blockRepository.GetBlocks(blockHashes);
+            List<Block> blocks = this.blockRepository.GetBlocks(chainedHeadersSnapshot.Select(c => c.Key).ToList());
```
Comment (dangershony) [inline src/Stratis.Features.FederatedPeg/TargetChain/CrossChainTransferStore.cs]: Generally we should not be calling `this.blockRepository` but we should get blocks via `ConsensusManager` so we look in cache first.

Probably not a fix for this PR.

**Comment by dangershony** [comment]
@quantumagi please review this part as well.

**Comment by quantumagi** [comment]
@fenix2222 , this seems to align with what we discussed. LGTM.

File: src/Stratis.Features.FederatedPeg/TargetChain/CrossChainTransferStore.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Features.FederatedPeg/TargetChain/CrossChainTransferStore.cs]: @dangershony you are right, I will make this change on his branch quick.

Comment (fassadlr) [reply src/Stratis.Features.FederatedPeg/TargetChain/CrossChainTransferStore.cs]: @dangershony actually no, this change affects tests etc. @fenix2222 Please make this change in another PR?

**Pull request merged**